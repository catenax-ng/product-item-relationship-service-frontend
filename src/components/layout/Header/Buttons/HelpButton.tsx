import { Button } from "cx-portal-shared-components";
import { useTranslation } from "react-i18next";

/**
 * Header Help Button
 * currently navigates to the help page in the github repository
 * @returns React component
 */
export const HelpButton = () => {
  const { t } = useTranslation();

  return (
    <Button
      color="secondary"
      size="small"
      sx={{
        backgroundColor: "white",
        marginRight: "16px",
      }}
      variant="contained"
      onClick={() => {
        window.open(import.meta.env.VITE_USER_MANUAL_URL, "_blank");
      }}
    >
      {t("content.irs.header.help")}
    </Button>
  );
};
