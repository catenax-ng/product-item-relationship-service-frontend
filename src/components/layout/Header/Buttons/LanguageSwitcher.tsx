import { LanguageSwitch } from "cx-portal-shared-components";
import { I18nService } from "../../../../lib";

export const LanguageSwitcher: React.FC = () => {
  return (
    <LanguageSwitch
      current={localStorage.getItem("i18nextLng") ?? "en"}
      languages={[{ key: "de" }, { key: "en" }]}
      onChange={(lang) => {
        I18nService.changeLanguage(lang);
      }}
    />
  );
};
