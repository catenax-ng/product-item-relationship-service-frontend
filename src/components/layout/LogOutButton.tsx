import { useKeycloak } from "@react-keycloak/web";
import { Button } from "cx-portal-shared-components";
import { redirect } from "react-router-dom";

export const LogOutButton: React.FC = () => {
  const { keycloak } = useKeycloak();
  return (
    <Button
      onClick={async () => {
        await keycloak.logout();
        redirect("/");
      }}
    >
      Log Out
    </Button>
  );
};
