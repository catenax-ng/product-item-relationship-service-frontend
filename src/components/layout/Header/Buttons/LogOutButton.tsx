import styled from "@emotion/styled";
import { useKeycloak } from "@react-keycloak/web";
import { UserAvatar, UserMenu, UserNav } from "cx-portal-shared-components";
import { useMemo, useRef, useState } from "react";
import { Link } from "react-router-dom";
import { LanguageSwitcher } from "./LanguageSwitcher";

/**
 * Header Log out Button
 * @returns React component
 */
export const LogOutButton: React.FC = () => {
  const { keycloak } = useKeycloak();
  const avatar = useRef<HTMLDivElement>(null);
  const [userInfo, setUserInfo] = useState<{ name: string; tenant: string } | undefined>();
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  useMemo(async () => {
    const info = (await keycloak.loadUserInfo()) as { name: string; tenant: string };
    setUserInfo(info);
  }, []);

  const openMenu = () => setIsMenuOpen(true);
  const onClickAway = (e: MouseEvent | TouchEvent) => {
    if (!avatar.current?.contains(e.target as HTMLDivElement)) {
      setIsMenuOpen(false);
    }
  };

  return (
    <StyledUserInfo>
      <div ref={avatar}>
        <UserAvatar onClick={openMenu} />
      </div>
      <UserMenu
        open={isMenuOpen}
        userName={userInfo?.name ?? ""}
        userRole={userInfo?.tenant ?? ""}
        onClickAway={onClickAway}
      >
        <UserNav
          component={Link}
          divider
          items={[
            {
              title: "Logout",
              to: "/",
              onClick: async () => {
                await keycloak.logout();
              },
            },
          ]}
        />
        <LanguageSwitcher />
      </UserMenu>
    </StyledUserInfo>
  );
};

const StyledUserInfo = styled.div`
  position: relative;
  z-index: 100;
`;
