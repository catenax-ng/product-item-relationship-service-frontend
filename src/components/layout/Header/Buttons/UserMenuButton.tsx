import styled from "@emotion/styled";
import { UserAvatar, UserMenu, UserNav } from "cx-portal-shared-components";
import { useRef, useState } from "react";
import { Link } from "react-router-dom";
import { useCustomKeycloak } from "../../../../lib/keycloak";
import { getUserInfo } from "../../../../services/queries/user";
import { LanguageSwitcher } from "./LanguageSwitcher";

/**
 * Header Log out Button
 * @returns React component
 */
export const UserMenuButton: React.FC = () => {
  const { logout } = useCustomKeycloak();
  const avatar = useRef<HTMLDivElement>(null);

  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const { data: userInfo } = getUserInfo();

  const openMenu = () => setIsMenuOpen(true);
  const onClickAway = (e: MouseEvent | TouchEvent) => {
    if (!avatar.current?.contains(e.target as HTMLDivElement)) {
      setIsMenuOpen(false);
    }
  };

  return (
    <StyledUserInfo>
      <div ref={avatar}>
        <UserAvatar onClick={openMenu} />
      </div>
      <UserMenu
        open={isMenuOpen}
        userName={userInfo?.name ?? ""}
        userRole={userInfo?.tenant ?? ""}
        onClickAway={onClickAway}
      >
        <UserNav
          component={Link}
          divider
          items={[
            {
              title: "Logout",
              to: "/",
              onClick: async () => {
                await logout();
              },
            },
          ]}
        />
        <LanguageSwitcher />
      </UserMenu>
    </StyledUserInfo>
  );
};

const StyledUserInfo = styled.div`
  position: relative;
  z-index: 100;
`;
