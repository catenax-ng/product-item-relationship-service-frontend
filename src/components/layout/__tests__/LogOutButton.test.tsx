import { vi } from "vitest";
import { fireEvent, renderWithRouter, screen } from "../../../testing/test-utils";
import { LogOutButton } from "../Header/Buttons/LogOutButton";

const logout = vi.fn();

vi.mock("@react-keycloak/web", () => {
  return {
    useKeycloak: () => ({
      keycloak: {
        logout: () => {
          logout();
        },
      },
    }),
  };
});

it("renders the logout button", () => {
  const { container } = renderWithRouter(<LogOutButton />);
  expect(container).toMatchSnapshot();
});

it("clicks the logout button", () => {
  renderWithRouter(<LogOutButton />);
  const button = screen.getByRole("button");
  expect(button).toBeDefined();
  fireEvent.click(button);
  expect(logout).toHaveBeenCalledTimes(1);
});
