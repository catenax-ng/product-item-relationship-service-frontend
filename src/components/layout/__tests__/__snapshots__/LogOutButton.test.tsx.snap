// Vitest Snapshot v1

exports[`renders the logout button 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LogoutIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
